File: leetcode/977.squares-of-a-sorted-array/main.go
package main

import "fmt"

func abs(a int) int {
	if a < 0 {
		return -a
	} else {
		return a
	}
}

func sortedSquares(nums []int) []int {
	n := len(nums)
	res := make([]int, n)
	l, r, p := 0, n-1, n-1
	for l <= r {
		if abs(nums[l]) > abs(nums[r]) {
			res[p] = nums[l] * nums[l]
			l++
		} else {
			res[p] = nums[r] * nums[r]
			r--
		}
		p--
	}
	return res
}

func main() {
	fmt.Println(sortedSquares([]int{-4, -1, 0, 2, 5}))
}

File: leetcode/977.squares-of-a-sorted-array/main.rs
pub struct Solution;

impl Solution {
    pub fn sorted_squares(nums: Vec<i32>) -> Vec<i32> {
        let n = nums.len();
        let mut arr = vec![0; n];
        let (mut i, mut j, mut pos) = (0, n, n)
        while i < j {
            pos -= 1;
            if nums[i].abs() > nums[j - 1].abs() {
                arr[pos] = nums[i] * nums[i];
                i += 1;
            } else {
                j -= 1;
                arr[pos] = nums[j] * nums[j];
            }
        }
        arr
    }
}

fn main() {
    println!("{:?}", Solution::sorted_squares(vec![-4, -1, 0, 3, 10]));
}
